0030-simple-classical-representations.md

```diff
@@ -0,0 +1,260 @@
+
+This is not intended to be a complete set of classical expressions, for simplicity in this initial implementation.
+Notably, this MVP will _not_ include:
+
+1. arithmetic on `uint[n]` (`+`, `*`, etc);
+2. indexing the bits resulting from one of these rvalue expresions;
+3. slicing a `uint[n]` to retrieve a subset of its bits;
+4. any method of assigning one of these expressions for reuse elsewhere in the circuit.
+
+Points 1 through 3 are rejected in this initial proposal for the sake of keeping the number of expression-tree nodes and type-system entries small.
+Point 4 is rejected because it would require the definition of additional `Operation` objects and potentially modifications to the `QuantumCircuit` class that we are not yet prepared to make.
+All of these points are eligible to be included in future work, just not the MVP.
+
+
+### Construction
+
+#### Stage 1
+
+This is a temporary representation, so it is a non-goal to produce the best possible UX for constructing these objects.
+In the initial representation phase, users will be required to build the expression-tree representations themselves.
+This will look _something_ like (names up for discussion):
+
+```python
+from qiskit.circuit import Clbit, ClassicalRegister
+from qiskit.circuit.classical import expr
+
+loose = [Clbit() for _ in [None]*3]
+cr1 = ClassicalRegister(2)
+cr2 = ClassicalRegister(2)
+
+# Equivalent to the current condition `(cr, 2)`:
+equal_reg = expr.Equal(expr.Value(cr1), expr.Value(2))
+
+# Equivalent to the current condition `(loose[0], False)`
+equal_bit = expr.Equal(expr.Value(loose[0]), expr.Value(False))
+
+# The above example, but using a Bool-valued unary expression:
+not_bit = expr.Not(expr.Value(loose[0]))
+
+# The comparison 0 < cr1 <= cr2
+bounded_reg = expr.And(
+    expr.Less(expr.Value(0), expr.Value(cr1),
+    expr.LessEqual(expr.Value(cr1), expr.Value(cr2),
+)
+```
+
+These expressions will then be given directly to `IfElseOp` or `WhileLoopOp` in their `condition` fields.
+These expressions will not be valid in the general `Instruction.condition` field; we do not want to expand any support for something we are already moving to remove.
+
+
+#### Stage 2
+
+Looking beyond the initial phase, it will be convenient to define a less verbose way for users to create these expressions.
+Since this document is intended to define a representation that can be removed and deprecated from Terra easily if it needs to evolve, we will _not_ consider overloading the Python magic-methods `__and__` etc on `Clbit` and `ClassicalRegister` in general Terra usage.
+However, we can do so in a limited scope, which will temporarily change the semantics.
+This will look like (continued from above code block, and the same examples in the same order):
```

Comment:
I think it's worth evaluating how much this makes the deprecation and eventual removal path smoother for users vs. the near time usability cost. It seems like most usage of `IfElseOp` and `WhileLoopOp` will want a `with expr.build():` block, and (with the exception of `__eq__`) these magic methods aren't currently being used, and so could be deprecated as needed when used.

Reply:
If we choose to keep these objects, we can also just move the magic methods onto the core objects, but that doesn't really feel great to me; I think magic methods should always feel _obvious_ about what they're going to do, and detached from a builder context, I don't think "`Clbit == Clbit` makes a runtime expression" is super obvious.